File: azureml/pipeline/wrapper/_widgets/_validation.py
# ---------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# ---------------------------------------------------------
from traitlets import Unicode

from ipywidgets import DOMWidget, register


@register
class ValidateView(DOMWidget):
    """ A widget for the pipeline validate visualization.

    Especially, the widget accept message to update running status.
    And return the latest status into view.
    """

    _view_name = Unicode('ValidateView').tag(sync=True)
    _view_module = Unicode('validate_widget').tag(sync=True)
    _view_module_version = Unicode('0.0.0').tag(sync=True)

    graph_json = Unicode().tag(sync=True)
    env_json = Unicode().tag(sync=True)
    lib_url = Unicode().tag(sync=True)
    container_id = Unicode().tag(sync=True)
    visualize_id = Unicode().tag(sync=True)

    def __init__(self, **kwargs):
        """Create a ValidateView widget."""
        super(DOMWidget, self).__init__(**kwargs)
        self.graph_json = kwargs.get("graph_json")
        self.env_json = kwargs.get("env_json")
        self.lib_url = kwargs.get("lib_url")
        self.container_id = kwargs.get("container_id")
        self.visualize_id = kwargs.get("visualize_id")

        def handle_msg(instance, message, params):
            if message.get("message") == "log_query:request":
                self.handle_log_query_request(message.get("body"))

        self.on_msg(handle_msg)

    def handle_log_query_request(self, body):
        import requests
        uid = body.get("uid")
        content = body.get("content")
        response = requests.get(content.get("url"))
        self.send_message("log_query:response", {
            "uid": uid,
            "result": response.text
        })

    def send_message(self, message: str, content: dict):
        self.send({
            "message": message,
            "body": content
        })

File: azureml/pipeline/wrapper/_widgets/js/validation.js
try {
    require.undef("validate_widget")

    define('validate_widget', ["@jupyter-widgets/base"], function(widgets) {
        var ValidateView = widgets.DOMWidgetView.extend({
            render () {
                window.widget_self = this
                var visualize_id = this.model.get('visualize_id')

                if (!window._renderLock) {
                    window._renderLock = {}
                }
                if (window._renderLock[visualize_id]) {
                    return
                }
                window._renderLock[visualize_id] = "widget"
                console.log("load as widget", Date.now())

                var lib_url = this.model.get('lib_url')
                var graph_json = JSON.parse(this.model.get('graph_json'))
                var env_json = JSON.parse(this.model.get('env_json'))
                var container_id = this.model.get('container_id')

                window.render_container_id = container_id
                window.graph_json = graph_json
                window.env_json = env_json
                window.before_script = performance.now()

                var container = document.createElement('div')
                container.id = container_id
                this.el.appendChild(container)

                var style = document.createElement('style')
                style.innerHTML = [
                    "#", container_id, " svg.react-dag-editor-svg-container { height: 800px; }",
                    ".cell-output-ipywidget-background { background: transparent !important }"
                ].join('')
                this.el.appendChild(style)

                this.model.on('msg:custom', dispatchMessage, this);

                if (!window.__event_hub) {
                    window.__event_hub = {}
                }
                if (!window.__event_hub[container_id]) {
                    window.__event_hub[container_id] = {}
                }

                if (!window.__send_event) {
                    window.__send_event = {}
                }
                window.__send_event[container_id] = sendMessage.bind(this)

                function sendMessage(message, uid, content) {
                    return new Promise((resolve) => {
                        this.model.send({
                            message: `${message}:request`,
                            body: {
                                uid,
                                content
                            }
                        })
    
                        var respMessageKey = `${message}:response`
                        if (!window.__event_hub[container_id][respMessageKey]) {
                            window.__event_hub[container_id][respMessageKey] = []
                        }
                        window.__event_hub[container_id][respMessageKey].push(callback)
    
                        function callback (response) {
                            if (response.uid !== uid) {
                                return
                            }

                            var idx = window.__event_hub[container_id][respMessageKey].indexOf(callback) 
                            window.__event_hub[container_id][respMessageKey].splice(idx, 1)
                            
                            resolve(response)
                        }
                    })
                }

                function dispatchMessage (rawMessage) {
                    var message = rawMessage.message
                    var body = rawMessage.body

                    if (!window.__event_hub[container_id][message]) {
                        window.__event_hub[container_id][message] = []
                    }
                    var listeners = window.__event_hub[container_id][message]

                    listeners.forEach(cb => {
                        try {
                            cb(body)
                        } catch (e) {
                            console.error("Unexpected error in listener", e)
                        }
                    })

                    console.log(body)
                }

                var script = document.createElement('script')
                script.src = lib_url
                this.el.appendChild(script)
            }
        });

        return {
            ValidateView
        }
    })
} catch (e) {
    console.log("create validation widget failed", e)
}